clio-client/src/main/scala/org/broadinstitute/clio/client/dispatch/DeleteExecutor.scala

```diff
@@ -1,132 +1,154 @@
-    ioUtil: IoUtil,
-    existingMetadata: deleteCommand.index.MetadataType
-  )(implicit ec: ExecutionContext): Future[UpsertId] = {
-    val pathsToDelete = existingMetadata.pathsToDelete.filter { path =>
-      val pathExists = ioUtil.googleObjectExists(path)
-      val err = s"'$path' associated with $prettyKey does not exist in the cloud."
-      if (!pathExists) {
-        if (!deleteCommand.force) {
-          throw new IllegalStateException(
-            err + " Use --force to mark the record as deleted in Clio"
+  private def checkPaths(
+    paths: immutable.Iterable[URI],
+    ioUtil: IoUtil
+  ): Source[immutable.Iterable[URI], NotUsed] = {
+    Source[URI](paths)
+      .mapAsync(paths.size) { path =>
+        if (!ioUtil.isGoogleObject(path)) {
+          Future.successful(
+            Left(
+              new IllegalStateException(
+                s"Inconsistent state detected, non-cloud path '$path' registered to $prettyKey"
+              )
+            )
           )
+        } else {
+          Future(Either.catchNonFatal(ioUtil.googleObjectExists(path))).map { attempt =>
+            attempt.flatMap { doesExist =>
+              if (doesExist) {
+                Right(Some(path))
+              } else {
+                val msg =
+                  s"'$path' associated with $prettyKey does not exist in the cloud."
+                if (deleteCommand.force) {
+                  logger.warn(msg)
+                  Right(None)
+                } else {
+                  Left(new IllegalStateException(msg))
+                }
+              }
+            }
+          }
         }
-        logger.warn(err)
       }
-      pathExists
-    }
-
-    val googleDeletes = pathsToDelete.map { path =>
-      Future {
-        logger.info(s"Deleting '$path' in the cloud.")
-        if (ioUtil.deleteGoogleObject(path) != 0) {
-          // We log immediately here because we Future.sequence these together later,
-          // and that operation short-circuits and returns the first error it hits, but
-          // we want messages for all failed deletes to appear in the logs.
-          val err = s"Could not delete '$path' in the cloud."
-          logger.error(err)
-          throw new Exception(err)
+      .fold(Right(Seq.empty[URI]): Either[Seq[Throwable], Seq[URI]]) { (acc, attempt) =>
+        (acc, attempt) match {
+          case (Right(ps), Right(maybePath)) => Right(ps ++ maybePath.toIterable)
+          case (Right(_), Left(err))         => Left(Seq(err))
+          case (Left(errs), Right(_))        => Left(errs)
+          case (Left(errs), Left(err))       => Left(errs :+ err)
         }
       }
-    }
+      .flatMapConcat {
+        _.fold(
+          errs => Source.failed(new IllegalStateException(errs.mkString(", "))),
+          paths => Source.single(paths.to[immutable.Iterable])
+        )
+      }
+  }
 
-    val markedAsDeleted = existingMetadata.markDeleted(deleteCommand.note)
+  private def deletePaths(
```

Comment:
This is also defined in the move executor, so it'd be nice to find a common spot for it. My gut says `IoUtil` but I'm not sure if that class should know about streams... what do people think?

Reply:
(It doesn't seem as weird to have in `IoUtil` if it was defined in terms of `IO` and `fs2.Stream` instead of `Future` and `Source`, so maybe I've just got a weird aversion to Akka proliferation)